### PR TITLE
wait_til_resource_exists

### PR DESCRIPTION
**Pull request by baniobloom**
### What does this PR do?
Added function wait_til_resource_exists to avoid errors when trying to modify a resource before it is fully available
### What issues does this PR fix or reference?
This bug: https://github.com/saltstack/salt/issues/46959#issuecomment-383838577

Added function wait_til_resource_exists to avoid errors when trying to modify a resource before it is fully available
### Previous Behavior
Creation of a vpc with subnets, route tables, internet gateway, vpc peering connection, etc would usually fail because the module boto_vpc.py would try to modify a resource before it was fully available.

### New Behavior
The function wait_til_resource_exists waits until the resource if fully available

### Tests written?

No

### Commits signed with GPG?

No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by rallytime** [comment]
HI @baniobloom - This change is causing the following tests to fail:

- unit.states.test_boto_vpc.BotoVpcInternetGatewayTestCase.test_present_when_resource_does_not_exist
- unit.states.test_boto_vpc.BotoVpcSubnetsTestCase.test_present_when_resource_does_not_exist

https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu16-py3/9819/

Can you take a look?

**Comment by rallytime** [comment]
re-run py

**Comment by rallytime** [comment]
re-run lint

File: salt/modules/boto_vpc.py [hunk 1]
```diff
@@ -487,6 +489,30 @@ def resource_exists(resource, name=None, resource_id=None, tags=None,
     except BotoServerError as e:
         return {'error': salt.utils.boto.get_error(e)}
 
+def wait_til_resource_exists(resource_type, resource_id, region=None, key=None, keyid=None, profile=None):
+    '''Wait for the resource to exist/be ready
```
Comment (rallytime) [inline salt/modules/boto_vpc.py]: Can you change this to be on two separate lines? The `'''` and the `Wait for ...` should be split up.

Can you also add a `.. versionadded:: 2017.7.7` directive to these docs as well as a CLI example?

**Review by rallytime — CHANGES_REQUESTED** [review]
Thanks for all of these fixes @baniobloom! I have some minor documentation requests on this, but otherwise this LGTM.

**Comment by baniobloom** [comment]
@rallytime updated.

**Review by rallytime — APPROVED** [review]

**Comment by cachedout** [comment]
@ryan-lane Do you have any feedback here?

**Comment by rallytime** [comment]
@baniobloom This has some lint violations. Can you fix them?

https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/22203/violations/file/salt/modules/boto_vpc.py/

**Comment by rallytime** [comment]
@baniobloom There are some related test failures with this change:

- unit.states.test_boto_vpc.BotoVpcSubnetsTestCase.test_present_when_resource_does_not_exist
- unit.states.test_boto_vpc.BotoVpcInternetGatewayTestCase.test_present_when_resource_does_not_exist

https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-cent7-py3/5361/

Can you take a look?

**Comment by rallytime** [comment]
Since we haven't heard back here for a while, I am going to close this. Please let us know when you're ready and we will gladly open this again! Or, feel free to open a new PR. Whichever is easiest for you. :)

Thanks!

**Comment by baniobloom** [comment]
@rallytime can't get to: https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-cent7-py3/5361/

**Comment by baniobloom** [comment]
@rallytime can you reopen?

**Comment by rallytime** [comment]
@baniobloom I've reopened and the tests should start again, but I posted the ones that are failing above so you can try them when the tests timeout. (That's what those 404 errors mean.)

**Comment by rallytime** [comment]
Here's the output from the failing tests in this run https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu16-py3/11191/:

- unit.states.test_boto_vpc.BotoVpcInternetGatewayTestCase.test_present_when_resource_does_not_exist 
```
Error Message

file descriptor cannot be a negative integer (-1)

Stacktrace

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/moto/core/models.py", line 71, in wrapper
    result = func(*args, **kwargs)
  File "/testing/tests/unit/states/test_boto_vpc.py", line 207, in test_present_when_resource_does_not_exist
    name='test', vpc_name='test', **self.extra_kwargs)
  File "/testing/salt/states/boto_vpc.py", line 809, in internet_gateway_present
    profile=profile)
  File "/testing/salt/modules/boto_vpc.py", line 1197, in create_internet_gateway
    wait_til_resource_exists(resource_type='internet_gateway', resource_id=r['id'], region=region, key=key, keyid=keyid, profile=profile)
  File "/testing/salt/modules/boto_vpc.py", line 517, in wait_til_resource_exists
    response = getattr(conn3, resource_id_argument_keyword_dict[resource_type]['command'])(**kwargs)
  File "/usr/local/lib/python3.5/dist-packages/botocore/client.py", line 314, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/local/lib/python3.5/dist-packages/botocore/client.py", line 599, in _make_api_call
    operation_model, request_dict)
  File "/usr/local/lib/python3.5/dist-packages/botocore/endpoint.py", line 148, in make_request
    return self._send_request(request_dict, operation_model)
  File "/usr/local/lib/python3.5/dist-packages/botocore/endpoint.py", line 177, in _send_request
    success_response, exception):
  File "/usr/local/lib/python3.5/dist-packages/botocore/endpoint.py", line 273, in _needs_retry
    caught_exception=caught_exception, request_dict=request_dict)
  File "/usr/local/lib/python3.5/dist-packages/botocore/hooks.py", line 227, in emit
    return self._emit(event_name, kwargs)
  File "/usr/local/lib/python3.5/dist-packages/botocore/hooks.py", line 210, in _emit
    response = handler(**kwargs)
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 183, in __call__
    if self._checker(attempts, response, caught_exception):
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 251, in __call__
    caught_exception)
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 269, in _should_retry
    return self._checker(attempt_number, response, caught_exception)
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 317, in __call__
    caught_exception)
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 223, in __call__
    attempt_number, caught_exception)
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 359, in _check_caught_exception
    raise caught_exception
  File "/usr/local/lib/python3.5/dist-packages/botocore/endpoint.py", line 222, in _get_response
    proxies=self.proxies, timeout=self.timeout)
  File "/usr/local/lib/python3.5/dist-packages/botocore/vendored/requests/sessions.py", line 573, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/botocore/vendored/requests/adapters.py", line 370, in send
    timeout=timeout
  File "/usr/local/lib/python3.5/dist-packages/botocore/vendored/requests/packages/urllib3/connectionpool.py", line 533, in urlopen
    conn = self._get_conn(timeout=pool_timeout)
  File "/usr/local/lib/python3.5/dist-packages/botocore/vendored/requests/packages/urllib3/connectionpool.py", line 237, in _get_conn
    if conn and is_connection_dropped(conn):
  File "/usr/local/lib/python3.5/dist-packages/botocore/vendored/requests/packages/urllib3/util/connection.py", line 39, in is_connection_dropped
    p.register(sock, POLLIN)
ValueError: file descriptor cannot be a negative integer (-1)
```
- unit.states.test_boto_vpc.BotoVpcSubnetsTestCase.test_present_when_resource_does_not_exist
```
Error Message

file descriptor cannot be a negative integer (-1)

Stacktrace

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/moto/core/models.py", line 71, in wrapper
    result = func(*args, **kwargs)
  File "/testing/tests/unit/states/test_boto_vpc.py", line 207, in test_present_when_resource_does_not_exist
    name='test', vpc_name='test', **self.extra_kwargs)
  File "/testing/salt/states/boto_vpc.py", line 616, in subnet_present
    profile=profile)
  File "/testing/salt/modules/boto_vpc.py", line 923, in create_subnet
    wait_til_resource_exists(resource_type='subnet', resource_id=subnet_object_dict['id'], region=region, key=key, keyid=keyid, profile=profile)
  File "/testing/salt/modules/boto_vpc.py", line 517, in wait_til_resource_exists
    response = getattr(conn3, resource_id_argument_keyword_dict[resource_type]['command'])(**kwargs)
  File "/usr/local/lib/python3.5/dist-packages/botocore/client.py", line 314, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/local/lib/python3.5/dist-packages/botocore/client.py", line 599, in _make_api_call
    operation_model, request_dict)
  File "/usr/local/lib/python3.5/dist-packages/botocore/endpoint.py", line 148, in make_request
    return self._send_request(request_dict, operation_model)
  File "/usr/local/lib/python3.5/dist-packages/botocore/endpoint.py", line 177, in _send_request
    success_response, exception):
  File "/usr/local/lib/python3.5/dist-packages/botocore/endpoint.py", line 273, in _needs_retry
    caught_exception=caught_exception, request_dict=request_dict)
  File "/usr/local/lib/python3.5/dist-packages/botocore/hooks.py", line 227, in emit
    return self._emit(event_name, kwargs)
  File "/usr/local/lib/python3.5/dist-packages/botocore/hooks.py", line 210, in _emit
    response = handler(**kwargs)
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 183, in __call__
    if self._checker(attempts, response, caught_exception):
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 251, in __call__
    caught_exception)
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 269, in _should_retry
    return self._checker(attempt_number, response, caught_exception)
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 317, in __call__
    caught_exception)
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 223, in __call__
    attempt_number, caught_exception)
  File "/usr/local/lib/python3.5/dist-packages/botocore/retryhandler.py", line 359, in _check_caught_exception
    raise caught_exception
  File "/usr/local/lib/python3.5/dist-packages/botocore/endpoint.py", line 222, in _get_response
    proxies=self.proxies, timeout=self.timeout)
  File "/usr/local/lib/python3.5/dist-packages/botocore/vendored/requests/sessions.py", line 573, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/botocore/vendored/requests/adapters.py", line 370, in send
    timeout=timeout
  File "/usr/local/lib/python3.5/dist-packages/botocore/vendored/requests/packages/urllib3/connectionpool.py", line 533, in urlopen
    conn = self._get_conn(timeout=pool_timeout)
  File "/usr/local/lib/python3.5/dist-packages/botocore/vendored/requests/packages/urllib3/connectionpool.py", line 237, in _get_conn
    if conn and is_connection_dropped(conn):
  File "/usr/local/lib/python3.5/dist-packages/botocore/vendored/requests/packages/urllib3/util/connection.py", line 39, in is_connection_dropped
    p.register(sock, POLLIN)
ValueError: file descriptor cannot be a negative integer (-1)
```
The test results time out after a couple of days, as we have quite the volume of PRs coming through. :)

I'm going to close this again for now until you have a change to fix the test failures. The above information should be enough to trouble shoot this, but do let me know if you have any other questions.

There's some documentation [here](https://docs.saltstack.com/en/latest/topics/development/tests/index.html) about running the test suite.

Also note that these failures are specific to Python 3. The Python 2 tests are not failing.

Once these tests are fixed up, we can certainly open this up again. :)

**Comment by baniobloom** [comment]
I'm not sure what this error is, but looks like some people are seeing it with older versions on moto.  When I run the boto_vpc unit test in my environment I don't see the error. Not sure if I'm doing it right:

`python3 tests/runtests.py -n unit.modules.test_boto_vpc`

What version of moto is the test environment using?  I'm using moto==1.3.3.

**Comment by rallytime** [comment]
@baniobloom We are also using moto==1.3.3. I included the boto and boto3 versions below as well, just in case that changes something:
```
20:16:42 ----------
20:16:42           ID: boto
20:16:42     Function: pip.installed
20:16:42         Name: boto >= 2.46.0
20:16:42       Result: True
20:16:42      Comment: All packages were successfully installed
20:16:42      Started: 20:12:05.382471
20:16:42     Duration: 4936.883 ms
20:16:42      Changes:   
20:16:42               ----------
20:16:42               boto==2.48.0:
20:16:42                   Installed
20:16:42 ----------
20:16:42           ID: boto3
20:16:42     Function: pip.installed
20:16:42       Result: True
20:16:42      Comment: All packages were successfully installed
20:16:42      Started: 20:12:10.320220
20:16:42     Duration: 12301.483 ms
20:16:42      Changes:   
20:16:42               ----------
20:16:42               boto3==1.7.49:
20:16:42                   Installed
20:16:42 ----------
20:16:42           ID: moto
20:16:42     Function: pip.installed
20:16:42       Result: True
20:16:42      Comment: All packages were successfully installed
20:16:42      Started: 20:12:22.622598
20:16:42     Duration: 21426.039 ms
20:16:42      Changes:   
20:16:42               ----------
20:16:42               moto==1.3.3:
20:16:42                   Installed
20:16:42 ----------
```

Sometimes the boto unit tests stomp on each other and if tests run before others, they affect the results. In those cases, we need to find where that is happening and fix up the issue.

What happens when your run all of the unit tests with Python 3? Maybe a test that is running earlier is affecting some results down the line.

```
python3 tests/runtests.py -u
```

**Pull request closed without merging**